source/extensions/filters/network/thrift_proxy/router/router_impl.cc

```diff
@@ -48,6 +48,37 @@ RouteEntryImplBase::RouteEntryImplBase(
   }
 }
 
+// Similar validation procedure with Envoy::Router::RouteEntryImplBase::validateCluster
+void RouteEntryImplBase::validateClusters(
+    const Upstream::ClusterManager::ClusterInfoMaps& cluster_info_maps) const {
+  // Currently, we verify that the cluster exists in the CM if we have an explicit cluster or
+  // weighted cluster rule. We obviously do not verify a cluster_header rule. This means that
+  // trying to use all CDS clusters with a static route table will not work. In the upcoming RDS
+  // change we will make it so that dynamically loaded route tables do *not* perform CM checks.
+  // In the future we might decide to also have a config option that turns off checks for static
+  // route tables. This would enable the all CDS with static route table case.
+  if (!cluster_name_.empty()) {
+    if (!cluster_info_maps.hasCluster(cluster_name_)) {
+      throw EnvoyException(fmt::format("route: unknown thrift cluster '{}'", cluster_name_));
+    }
+  } else if (!weighted_clusters_.empty()) {
+    for (const WeightedClusterEntrySharedPtr& cluster : weighted_clusters_) {
+      if (!cluster->clusterName().empty()) {
+        if (!cluster_info_maps.hasCluster(cluster->clusterName())) {
+          throw EnvoyException(
+              fmt::format("route: unknown thrift weighted cluster '{}'", cluster->clusterName()));
+        }
+      }
+      // For weighted clusters with `cluster_header_name`, we only verify that this field is
+      // not empty because the cluster name is not set yet at config time (hence the validation
+      // here).
+      else if (cluster->clusterHeader().get().empty()) {
+        throw EnvoyException("route: unknown thrift weighted cluster with no cluster_header field");
```

Comment:
Is there a test for this case?

Reply:
Good catch! Will add one in next commit. Thanks.

Reply:
After I took another look, actually it's a dead code (facepalm)
In http world, we can specified `cluster_header` filed inside the weighted clusters
https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/api/envoy/config/route/v3/route_components.proto#L361

Hence it make senses to check either we can validate the cluster or we have a cluster_header.

However, in thrift world. We don't have `cluster_header` for weighted clusters.
From the proto, we can see it's one of the three. We can't get both header and weighted cluster at the moment.
https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/api/envoy/config/route/v3/route_components.proto#L361

The `WeightedCluster::ClusterHeader` is propagated from its parent, which doesn't make senses. (note that `ClusterHeader` is one of the interface `RouteEntry` but I don't want to deviate too much at this moment)

To conclude, I remove this dead code given the proto sanitizer will do it for me 
also remove the empty check for the same reason
https://github.com/envoyproxy/envoy/blob/e4eaf1b970876d269f7e15dc049f90c79742e298/api/envoy/extensions/filters/network/thrift_proxy/v3/route.proto#L117

And add a test to ensure the coverage